---
title: "RNAseq counts processing for gene-level translational efficiency analysis"
output: html_notebook
---

This notebook describes the steps taken to process the gene-level RNA-seq counts for canonical CDS regions in the Hofman et al. paper (Table S1C). These counts were used for gene-level translational efficiency analyses of canonical genes (Figures 1G and 1H).

```{r setup, include = FALSE}
library(DESeq2)
library(tidyverse)
library(ggrepel)
```

# Process counts data
```{r}
# Load counts data
medullo_canonical_RNA_counts_TE <- read.table(
  "/rnaseq_for_te/data/processed/featureCounts/Medullo_RNA_GRCh38.102.geneCountsCDS.txt", 
  check.names = F, 
  header = T, 
  row.names = 1) %>%
  select(6:ncol(.))  # Remove irrelevant metadata columns

colnames(medullo_canonical_RNA_counts_TE) <- sapply(basename(colnames(medullo_canonical_RNA_counts_TE)),
                                                 function(x)
                                                   gsub(".Aligned.sortedByCoord.out.bam", "", x))

colnames(medullo_canonical_RNA_counts_TE) <- sapply(colnames(medullo_canonical_RNA_counts_TE), function(x) paste0(str_split(x, pattern = "_")[[1]][1:2], collapse = "_"))

write.csv(medullo_canonical_RNA_counts_TE, "/rnaseq_for_te/results/S1_RNA_for_te_canonical_counts.csv", quote = F)
```

