---
title: "MBL tissues RNAseq clustering (based on molecular subgroup)"
output: html_notebook
---

# Introduction

This notebook focuses on the analysis of RNA sequencing data from Medulloblastoma tissue samples. The goal is to identify differentially expressed genes and perform clustering based on MYC expression levels. The notebook involves loading necessary libraries, data preprocessing, running DESeq2 analysis, PCA visualization, identifying cluster-defining genes, and creating heatmaps.

Inputs: 
- Full length canonical RNA-seq reads (Table S1B)
- Sample metadata (Table S1A)

Outputs:
- Figure S1K

# 1. Setup and Libraries

First, we load the required libraries for the analysis: stringr, tidyverse, DESeq2, RColorBrewer, fgsea, msigdb, ExperimentHub, GSEABase, org.Hs.eg.db, AnnotationDbi, and egg.

```{r setup, include = FALSE}
library(stringr)
library(tidyverse)
library(DESeq2)
library(RColorBrewer)
library(fgsea)
library(msigdb)
library(ExperimentHub)
library(GSEABase)
library(org.Hs.eg.db)
library(AnnotationDbi)
library(egg)
library(ggbeeswarm)
```

# 2. Load and pre-process Data

## 2.1. Load and pre-process RNA-seq counts

We load the RNAseq counts table and fix sample column names. We also remove unnecessary metadata columns and write the processed data to a CSV file for further analysis.

```{r}
# Load RNAseq counts table (Table S1B)
canonical_RNA_counts_all <- read.table("/rnaseq_regular/data/processed/featurecounts/medullo_allsamples.counts", check.names = F)

# Fix sample column names
colnames(canonical_RNA_counts_all) <- sapply(basename(colnames(canonical_RNA_counts_all)),
                                             function(x)
                                               gsub(".Aligned.sortedByCoord.out.bam", "", x)) 

# Remove unnecessary metadata columns
canonical_RNA_counts_all <- canonical_RNA_counts_all %>%
  dplyr::select(6:ncol(.))

```

## 2.2. Load and process metadata

We load the metadata table associated with the RNA sequencing data. We filter samples with both RNAseq and riboseq data and select metadata for tissue samples. We also convert specific columns to factors and set rownames for easy indexing.

```{r}
# Load metadata
sample_metadata <- readxl::read_xlsx("Table_S1.xlsx", sheet = 1, col_names = T, skip = 2, .name_repair = "universal")

# Filter samples with both RNAseq and riboseq data
sample_metadata_complete <- sample_metadata[which(
  !is.na(sample_metadata$RNASeq_file_names) &
    !is.na(sample_metadata$RiboSeq_file_names..final.)), ]

# Extract RNA sample names based on filenames
sample_metadata_complete <- sample_metadata_complete %>%
  mutate(rna_name = gsub("_R1.*", "", gsub(";.*", "", .$RNASeq_file_names)))

# Select metadata for cell line samples
sample_metadata_tissues <- subset(sample_metadata_complete, Sample_Type != "Cell line")
sample_metadata_tissues$MYC.group <- as.factor(sample_metadata_tissues$MYC.group)

sample_metadata_tissues$G3G4_v_rest <- as.factor(sample_metadata_tissues$G3G4_v_rest)

rownames(sample_metadata_tissues) <- sample_metadata_tissues$Sample_ID
```

```{r}
# Exclude samples

# Tissue 10 is degraded; Tissue 12, Tissue 1, and Tissue 2 high immune infiltration. Exclude all autopsies because they are also of low quality. Remove tissue 6 because it is NOS.

sample_metadata_tissues <- sample_metadata_tissues %>%
  filter(!Sample_ID %in% c("Tissue_10", "Tissue_1", "Tissue_2", "Tissue_12")) %>%
  filter(!grepl("Autopsy", Sample_ID))

```

## 2.3. Match RNA-seq counts and metadata

```{r}
# Match RNA-seq counts columns to metadata rows
match_counts <- unique(str_extract(sample_metadata_tissues$RNASeq_file_names, str_c(colnames(canonical_RNA_counts_all), collapse = "|")))
canonical_RNA_counts <- canonical_RNA_counts_all[, match_counts]

# Find matching metadata
metadata_matches <- grep(paste(match_counts, collapse = "|"), sample_metadata_tissues$RNASeq_file_names)
sample_metadata_tissues <- sample_metadata_tissues[metadata_matches, ]
colnames(canonical_RNA_counts) <- sample_metadata_tissues$Sample_ID

sample_metadata_tissues$Molecular.subgroup <- as.factor(sample_metadata_tissues$Molecular.subgroup)
```

# 3. Differential expression analysis

## 3.1. Run DESeq2
```{r}
# Run DESeq2
dds <- DESeqDataSetFromMatrix(countData = canonical_RNA_counts, colData = sample_metadata_tissues, design = ~ G3G4_v_rest)
dds <- DESeq(dds)
vsd <- vst(dds, blind = F)
mat_vsd <- assay(vsd)
mat_dds <- counts(dds, normalized = T)

MYC_vsd <- mat_vsd["ENSG00000136997", ]
MYC_dds <- mat_dds["ENSG00000136997", ]

```


# 4. Sample clustering
This section of code is used to generate the plot for Figure S1K

```{r}
# Perform PCA and prepare data for plotting
pcaData <- plotPCA(vsd, intgroup=c("G3G4_v_rest"), returnData=TRUE)
pcaData$Sample_ID <- pcaData$name
pcaData$MYClvl <- MYC_vsd

# Define colors for PCA plots
col = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(27)

percentVar <- round(100 * attr(pcaData, "percentVar"))
```


```{r}
plt <- ggplot(pcaData, aes(PC1, PC2, color=G3G4_v_rest)) +
  scale_color_brewer(palette = "Set1") +
  geom_point(size=2) +
  ggrepel::geom_text_repel(aes(label = pcaData$name), vjust = 1, hjust = 1) +  # This line adds the labels
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
                                                           panel.background = element_blank(), axis.line = element_line(colour = "black"))+   
  theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5))
  # labs(color = "Subgroup", title = "RNA-seq")

plt

# ggsave("/rnaseq_regular/results/plots/RNA_PCA_tissues_subgroups.pdf")

```


# 5. Identify cluster-defining genes

We identify cluster-defining genes by analyzing the loadings of principal components. We sort the loadings for PC1 and PC2 and retrieve the gene symbols using the org.Hs.eg.db package.

```{r}
# Identify and select top varying genes
topVarGenes <- head(order(-rowVars(assay(vsd))),500)
mat_vsd_filtered <- mat_vsd[topVarGenes, ]

# Calculate PCA using base R function `prcomp`
pca <- prcomp(t(mat_vsd_filtered), center = TRUE, scale. = TRUE)

# The loadings are stored in the `rotation` component of the PCA object
loadings <- pca$rotation
```

```{r}
# Sort the loadings for the first principal component
PC1_loadings <- sort(loadings[, 1], decreasing = TRUE)

# The names of the genes with the highest loadings for PC1 are:
library(org.Hs.eg.db)
library(AnnotationDbi)
head(mapIds(x = org.Hs.eg.db,keys =  names(PC1_loadings), column = "SYMBOL", keytype = "ENSEMBL", multiVals = "first"), 20)

```

```{r}
# Sort the loadings for the first principal component
PC2_loadings <- sort(loadings[, 2], decreasing = TRUE)

# The names of the genes with the highest loadings for PC1 are:
library(org.Hs.eg.db)
library(AnnotationDbi)
head(mapIds(x = org.Hs.eg.db,keys =  names(PC2_loadings), column = "SYMBOL", keytype = "ENSEMBL", multiVals = "first"), 20)
head(names(PC2_loadings), 20)
```

## 5.1. PC1 and PC2 genes heatmaps

```{r}
library(gplots)

# Get the top 40 genes for PC1
top_genes_PC1 <- rownames(loadings)[order(abs(loadings[,1]), decreasing = TRUE)][1:40]

# Subset the expression matrix
subset_mat_vsd <- mat_vsd_filtered[top_genes_PC1,]

# Compute the Z-scores for better visualization
z_scores <- scale(t(subset_mat_vsd))

colnames(z_scores) <- mapIds(x = org.Hs.eg.db,keys =  colnames(z_scores), column = "SYMBOL", keytype = "ENSEMBL", multiVals = "first")

# Create the heatmap
heatmap.2(z_scores, 
          scale = "none", 
          Colv = NA, 
          trace = "none", 
          margin = c(8,6), 
          srtRow = 0, 
          srtCol = 90, 
          key.xlab = "Z-score", 
          main = "Heatmap of top genes on PC1")
```

```{r}
# Get the top 40 genes for PC2
top_genes_PC2 <- rownames(loadings)[order(abs(loadings[,2]), decreasing = TRUE)][1:40]

# Subset the expression matrix
subset_mat_vsd <- mat_vsd_filtered[top_genes_PC2,]

# Compute the Z-scores for better visualization
z_scores <- scale(t(subset_mat_vsd))

colnames(z_scores) <- mapIds(x = org.Hs.eg.db,keys =  colnames(z_scores), column = "SYMBOL", keytype = "ENSEMBL", multiVals = "first")

# Create the heatmap
heatmap.2(z_scores, 
          scale = "none", 
          Colv = NA, 
          trace = "none", 
          margin = c(8,6), 
          srtRow = 0, 
          srtCol = 90, 
          key.xlab = "Z-score", 
          main = "Heatmap of top genes on PC2")
```



```{r}
plt <- plotCounts(dds = dds, gene = "ENSG00000286053", intgroup = "G3G4_v_rest", returnData = T, normalized = T)

plt <- ggplot(data = plt, aes(y = count, x = G3G4_v_rest, color = G3G4_v_rest)) +
  scale_color_brewer(palette = "Set1") +
  labs(x = NULL, y = "ASNSD1-uORF RNA-seq counts") +
  geom_beeswarm() +
  theme_classic()
plt
```

```{r, fig.height = 4, fig.width = 4}
plt <- plotCounts(dds = dds, gene = "ENSG00000136997", intgroup = "G3G4_v_rest", returnData = T, normalized = T)

plt <- ggplot(data = plt, aes(y = count, x = G3G4_v_rest, color = G3G4_v_rest)) +
  scale_color_brewer(palette = "Set1") +
  labs(x = NULL, y = "MYC RNA-seq counts") +
  geom_beeswarm() +
  theme_classic()
plt
```

```{r}
plt <- plotCounts(dds = dds, gene = "ENSG00000136997", intgroup = "Molecular.subgroup", returnData = T, normalized = T)

plt <- ggplot(data = plt, aes(y = count, x = Molecular.subgroup, color = Molecular.subgroup)) +
  scale_color_brewer(palette = "Set1") +
  labs(x = NULL, y = "MYC RNA-seq counts") +
  geom_beeswarm() +
  theme_classic()
plt
```

```{r, fig.height = 4, fig.width = 4}
plt <- plotCounts(dds = dds, gene = "ENSG00000134323", intgroup = "G3G4_v_rest", returnData = T, normalized = T)

plt <- ggplot(data = plt, aes(y = count, x = G3G4_v_rest, color = G3G4_v_rest)) +
  scale_color_brewer(palette = "Set1") +
  labs(x = NULL, y = "MYCN RNA-seq counts") +
  geom_beeswarm() +
  theme_classic()
plt
```
```{r, fig.height = 4, fig.width = 5}
plt <- plotCounts(dds = dds, gene = "ENSG00000134323", intgroup = "Molecular.subgroup", returnData = T, normalized = T)

plt <- ggplot(data = plt, aes(y = count, x = Molecular.subgroup, color = Molecular.subgroup)) +
  scale_color_brewer(palette = "Set1") +
  labs(x = NULL, y = "MYCN RNA-seq counts") +
  geom_beeswarm() +
  theme_classic()
plt
```

